src/bin/coldsnap/main.rs

```diff
@@ -266,6 +268,27 @@ struct DownloadArgs {
     no_progress: bool,
 }
 
+
+/// Turn a user-specified tag string into a Tag object. Tags must start with 'KEY=' and
+/// denote the tag value with ',Value='.
+fn tag_from_str(input: &str) -> std::result::Result<Tag, String> {
+    const KEY_DELIMITER: &str = "Key=";
+    const VALUE_DELIMITER: &str = ",Value=";
+
+    if !input.starts_with(KEY_DELIMITER) {
+        return Err(format!("Tag inputs must start with '{KEY_DELIMITER}'").to_string());
+    }
+
+    if !input.contains(VALUE_DELIMITER) {
+        return Err(format!("Tag inputs must contain value entry with '{VALUE_DELIMITER}'").to_string());
+    }
+
+    //We have already validated the input contains both delimeters so there is no panic risk for unwrapping directly.
+    let k_v: (&str, &str) = input.split_once(KEY_DELIMITER).unwrap().1.split_once(VALUE_DELIMITER).unwrap();
+
+    Ok(Tag::builder().key(k_v.0).value(k_v.1).build())
```

Comment:
We should reject empty keys, since they lead to an error otherwise:
```
❯ cargo run -- upload --tag Key=,Value= --tag Key=A,Value=B Cargo.toml
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/coldsnap upload --tag Key=,Value= --tag Key=A,Value=B Cargo.toml`
Failed to upload snapshot: Failed to start snapshot: service error: ValidationException: Tag key cannot be null or empty.: ValidationException: Tag key cannot be null or empty.
```

```suggestion
    let (key, val) = input.split_once(KEY_DELIMITER).unwrap().1.split_once(VALUE_DELIMITER).unwrap();
    if key.is_empty() {
         return Err("Tag inputs must contain a non-empty key entry".to_string());
    }

    Ok(Tag::builder().key(key).value(val).build())
```

Reply:
Nice catch, I did miss this test above. I'm fine to make the change and do we want to do any of the other validations ourselves? The error spat was from letting EBS reject tag inputs so we don't have to lock step any changes to valid inputs over time. I'm letting the duplicate key and character limit fall through as well. The duplicate key would have to be done outside of the parsing function though.

Reply:
> Do we want to do any of the other validations ourselves?

I'm fine with it like this - seems unnecessary to repeat all the validations that the AWS SDK is going to perform. Even the non-empty key check is a tad excessive, but it's cheap to do and provides a better error message.